### ISSUE TITLE
Multiple peaks per cluster

### ISSUE THREAD

**Issue by miykael** [miykael/mni_atlas_reader#23]
Should we extract multiple peaks per cluster, like in SPM?

An idea would be to extract all peaks per cluster (ordered by intensity) but exclude new peaks if they are too close to already included peaks.

**Comment by danjgale** [comment]
+1 I like this idea. Sometimes one big cluster will have peaks in different ROIs and simply increasing the threshold doesn't solve this (as is often the case with clusters near M1 and SSc, for instance)

**Comment by miykael** [comment]
Exactly. I haven't used it, but I was told that `nilearn`'s `thresholding.get_clusters_table`, can do the extraction of the peaks within a cluster.

Here's the link to this: https://nistats.github.io/modules/generated/nistats.thresholding.get_clusters_table.html

@danjgale - do you want to work on that? Otherwise I can also work on that.

**Comment by miykael** [comment]
I tried out some different toolboxes (`nistats` and `nilearn`) but I haven't found yet a nice program that can read out multiple peaks per cluster. Perhaps in `scipy.ndiamge` and otherwise we could look at the code in SPM itself.
Does one of you have any idea or know a nice script to do this?

**Comment by rmarkello** [comment]
I think something like [this](http://scikit-image.org/docs/dev/auto_examples/segmentation/plot_peak_local_max.html) from `scikit-image` would work. It adds a whole other set of dependencies, so perhaps we can look for something a bit more lightweight (and/or the source code and include just a single function from the package with appropriate attribution?).

One of my questions would be how we clarify that the peaks are from the same cluster? The same `cluster_id` I guess?

**Comment by miykael** [comment]
True, I looked at the `scikit-image` thing as well, but I think this works only for 2D arrays, no?

SPM's sourcode might be the best approach - or we drop this issue for the moment.

About peaks per cluster - yes, I would just run the `find_peaks` routine for each cluster individually.

**Comment by danjgale** [comment]
Looking at SPMs code might be the best option, yeah

**Comment by miykael** [comment]
Ah, now I remember why this doesn't work... I tried the whole thing with an example stat_map:

```python
>>> import nibabel as nbbl
>>> data = nbbl.load('stat_map.nii.gz').get_data()
>>> peak_local_max(data, min_distance=10)
array([[182, 218, 182],
       [182, 218, 181],
       [182, 218, 180],
       ...,
       [ 10,  10,  12],
       [ 10,  10,  11],
       [ 10,  10,  10]])
>>> peak_local_max(data, min_distance=10).shape
(5871074, 3)
```

**Comment by rmarkello** [comment]
I think there is a way to do this with the help of clustering! You can provide `labels` to `peak_local_max()` so that each cluster as specified in `labels`, obtained from `scipy.ndimage.label()`, is considered separately. There might have to be some cleaning up because `peak_local_max()` returns _all_ the indices for flat clusters (of which most statistical maps have many!), but I think we can do that by calculating some distance metrics for outliers.

Let me give this a whack over the next couple of days. I am confident we can make this work, and I think it would be super helpful!

**Comment by rmarkello** [comment]
Wanted to send a minor update that I've got a working solution!!! 🎉 I'm just sorting out the best way to integrate it into the codebase, now, so be on the lookout for a PR.

I also found a small bug in the process, so hurrah for bug fixes, too?
